### PR TITLE
Server task5

### PR DESCRIPTION
**Pull request by dziq45**
This PR includes error handling and validation from server-task4.

### CODE REVIEW

File: server/middlewares/errorHandler.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+const {
+    NotFoundError,
+    UnauthorizedError,
+    AuthenticationError,
+    UnavailableServiceError,
+    BadRequestError } = require('../errorTypes/errorTypes')
+
+module.exports = handleError = (err, req, res, next) => {
+    switch (err.constructor) {
```
Comment (time2funk) [inline server/middlewares/errorHandler.js:9]: no this is not right.
please read the task description carefully.
_implement middleware to extend res with method handleError(error)_
you need to extend res object with method handleError(error)

File: server/controller/taskController.js [hunk 2]
```diff
@@ -1,59 +1,65 @@
...
                 res.status(200).json({ result });
             })
             .catch(err => {
-                res.status(500).json({ message: 'Something went wrong.' });
+                next(err);
+                //res.status(500).json({ message: 'Something went wrong.' });
             })
     },
-    addTask: (req, res) => {
+    addTask: (req, res, next) => {
```
Comment (time2funk) [inline server/controller/taskController.js:24]: please remove all next parameters from your controllers.
you need to use res.handleError(error)

File: server/middlewares/errorHandler.js (hunk 1 above)
Comment (dziq45) [reply server/middlewares/errorHandler.js]: Okay, I didn't understand how that should work, but I get it now and changed these next command to handleError

File: server/middlewares/errorHandler.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                    message:err.message
+                })
+            case UnavailableServiceError:
+                return res.status(503).json({
+                    message:err.message
+                })
+            case BadRequestError:
+                return res.status(400).json({
+                    message:err.message
+                })
```
Comment (time2funk) [inline server/middlewares/errorHandler.js:31]: please add default case 
err.message || 'Something went wrong'

File: server/middlewares/errorHandler.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            case UnauthorizedError:
+                return res.status(401).json({
+                    message:err.message
+                })
+            case AuthenticationError:
+                console.log('auth failed')
+                return res.status(403).json({
+                    message:err.message
```
Comment (time2funk) [inline server/middlewares/errorHandler.js:22]: I think it is better to have some predefined messages
message: err.message || 'Authentication failed'

File: server/middlewares/errorHandler.js (hunk 3 above)
Comment (dziq45) [reply server/middlewares/errorHandler.js:34]: Done

File: server/middlewares/errorHandler.js (hunk 4 above)
Comment (dziq45) [reply server/middlewares/errorHandler.js]: Do you mean messages stored in some file like:
`const AUTHENTICATION_ERROR = 'Authentication failed'`
or just messages displayed when err.message is not defined like in your example?

Added default message to every type of error.

File: server/service/validator.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+        const modelKeys = Object.keys(modelAttributes);
+        const filteredAttributes = Object.keys(attributes)
+            .filter(key => modelKeys.includes(key))
+            .reduce((acc, key) => Object.assign(acc, { [key]: attributes[key] }), {});
+
+
+        //check required
+        modelKeys.forEach(attributeKey => {
+            if (modelAttributes[attributeKey].required && filteredAttributes[attributeKey] == undefined) {
```
Comment (time2funk) [inline server/service/validator.js:14]: better to change filteredAttributes[attributeKey] == undefined 
to !filteredAttributes[attributeKey]

File: server/service/validator.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                throw new BadRequestError(`${attributeKey} must be provided`)
+            }
+        });
+
+        //check types
+        Object.keys(filteredAttributes).forEach(key => {
+            const constructorName = filteredAttributes[key].constructor.name
+            if (constructorName.toLowerCase() !== modelAttributes[key].type) {
```
Comment (time2funk) [inline server/service/validator.js:22]: if you wand to do that this way, then you need update all schemes to have proper value in type parameter.
for example 'string' -> 'String'

File: server/service/validator.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            }
+        })
+        //length, in
+        modelKeys.forEach(modelKey => {
+            const attribute = filteredAttributes[modelKey];
+            if (attribute) {
+                const rulesKeys = Object.keys(modelAttributes[modelKey])
+                rulesKeys.forEach(ruleKey => {
+                    const ruleValue = modelAttributes[modelKey][ruleKey];
+                    switch (ruleKey) {
```
Comment (time2funk) [inline server/service/validator.js:33]: please mode this rule checker into separate function and use it here

Comment (dziq45) [reply server/service/validator.js:35]: Can i make it a class then? Cause I can't use this .validateRules() in this module.exports

File: server/service/validator.js (hunk 5 above)
Comment (dziq45) [reply server/service/validator.js]: Done

File: server/service/validator.js (hunk 6 above)
Comment (dziq45) [reply server/service/validator.js]: Updated the types names and deleted toLowerCase()

**Pull request merged**